File: apps/api/core/session.py
"""
Session cookie sealing/unsealing for the 876 platform.

The session cookie is an HMAC-SHA256-signed JSON payload. It carries the
minimum data needed by Next.js proxy.ts to make routing decisions without
a round-trip to the API on every request:
  - userId, email, accessToken, exp

The Python API sets this cookie on every auth-completing response (login,
register, callback, verify-email, magic-otp-verify). The Next.js apps read
it from the browser's Cookie header and validate the HMAC locally using the
shared SESSION_COOKIE_SECRET.
"""

from __future__ import annotations

import base64
import hashlib
import hmac
import json
import time
from typing import Any


def account_identity(
    user_data: dict[str, Any],
    *,
    realm: str = "consumer",
    org_id: str | None = None,
    cross_realm: bool = False,
) -> dict[str, Any]:
    """Lightweight, token-free identity fields for one account.

    Shared by the active-account snapshot (top-level cookie fields) and each
    entry in the multi-account ``accounts`` list, so the chooser can display
    every signed-in account and ``switch_active`` can promote any of them.

    ``cross_realm`` is the per-user realm-gate exception, surfaced as
    ``crossRealm`` so the app proxies/guards can let an owner/super-admin into
    any realm. Carried on every account entry so switching accounts preserves it.
    """
    result: dict[str, Any] = {
        "userId": user_data.get("id") or user_data.get("userId") or "",
        "email": user_data.get("email", ""),
        "firstName": user_data.get("firstName") or user_data.get("first_name"),
        "lastName": user_data.get("lastName") or user_data.get("last_name"),
        "emailVerified": user_data.get("emailVerified") or user_data.get("email_verified", False),
        "avatar": user_data.get("avatar"),
        "username": user_data.get("username"),
        "realm": realm,
    }
    if org_id:
        result["orgId"] = org_id
    if cross_realm:
        result["crossRealm"] = True
    return result


def account_entry(
    user_data: dict[str, Any],
    session_id: str,
    *,
    realm: str = "consumer",
    org_id: str | None = None,
    cross_realm: bool = False,
) -> dict[str, Any]:
    """One entry in the multi-account ``accounts`` list — identity + its sid."""
    return {
        **account_identity(user_data, realm=realm, org_id=org_id, cross_realm=cross_realm),
        "sid": session_id,
    }


def merge_accounts(existing: list[dict[str, Any]] | None, new_entry: dict[str, Any]) -> list[dict[str, Any]]:
    """Add ``new_entry`` to the signed-in account set, de-duped by ``userId``.

    Re-authenticating an account already in the set replaces its stale entry
    (and sid) rather than appending a duplicate. Order is preserved with the
    freshly-authenticated account last.
    """
    user_id = new_entry.get("userId")
    kept = [a for a in (existing or []) if a.get("userId") != user_id]
    return [*kept, new_entry]


def select_account(accounts: list[dict[str, Any]] | None, sid: str) -> dict[str, Any] | None:
    """Return the account entry whose ``sid`` matches, or None when absent.

    The caller MUST use this to confirm a switch target belongs to the cookie's
    own account set — never trust a sid supplied by the client directly.
    """
    for account in accounts or []:
        if account.get("sid") == sid:
            return account
    return None


def seal_session(
    user_data: dict[str, Any],
    access_token: str | None,
    secret: str,
    ttl_seconds: int = 60 * 60 * 24 * 400,
    session_id: str | None = None,
    accounts: list[dict[str, Any]] | None = None,
    realm: str = "consumer",
    org_id: str | None = None,
    cross_realm: bool = False,
) -> str:
    """Return a signed, base64-encoded session cookie value.

    The active account's identity stays at the top level (unchanged contract,
    so existing proxy/session readers keep working). When ``session_id`` and
    ``accounts`` are provided, the cookie additionally carries the active
    session id (``sid``) and the full signed-in account set for multi-account
    switching.
    """
    payload_dict: dict[str, Any] = {
        **account_identity(user_data, realm=realm, org_id=org_id, cross_realm=cross_realm),
        "accessToken": access_token,
        "exp": int(time.time()) + ttl_seconds,
    }
    if session_id is not None:
        payload_dict["sid"] = session_id
    if accounts is not None:
        payload_dict["accounts"] = accounts
    payload = json.dumps(payload_dict, separators=(",", ":"))
    sig = _sign(payload, secret)
    # Unpadded base64url — avoids a trailing '=' in the cookie value, which some
    # cookie parsers/middleware mishandle.
    raw = base64.urlsafe_b64encode(f"{payload}.{sig}".encode()).decode()
    return raw.rstrip("=")


def unseal_session(cookie_value: str, secret: str) -> dict[str, Any] | None:
    """Verify the signature and return the session payload, or None on failure."""
    try:
        # Restore base64 padding stripped by seal_session.
        padded = cookie_value + "=" * (-len(cookie_value) % 4)
        decoded = base64.urlsafe_b64decode(padded.encode()).decode()
        # Split on last dot to separate payload from signature
        last_dot = decoded.rfind(".")
        if last_dot == -1:
            return None
        payload_str = decoded[:last_dot]
        sig = decoded[last_dot + 1 :]

        expected = _sign(payload_str, secret)
        if not hmac.compare_digest(sig, expected):
            return None

        payload = json.loads(payload_str)
        if not isinstance(payload, dict):
            return None
        if payload.get("exp", 0) < time.time():
            return None
        return payload
    except Exception:
        return None


def _sign(payload: str, secret: str) -> str:
    return hmac.new(secret.encode("utf-8"), payload.encode("utf-8"), hashlib.sha256).hexdigest()

File: apps/api/src/platform/session.ts
/**
 * Session cookie sealing and unsealing.
 *
 * The cookie is an HMAC-SHA256-signed JSON payload carrying the minimum the
 * Next.js apps need to make routing decisions without a round-trip to the API:
 * the active account's identity, its access token, and an expiry.
 *
 * ## The wire format is a cross-service contract
 *
 * ```
 * base64url( <compact JSON payload> "." <hex HMAC-SHA256(payload, secret)> )
 * ```
 *
 * Unpadded — a trailing `=` is stripped, because some cookie parsers and
 * middleware mishandle it. `@876/core`'s `verifySession876` is the reader on the
 * app side and only ever verifies; this module is the only writer. Both sides
 * recompute the HMAC over the decoded payload string, so a cookie sealed here
 * verifies there byte for byte — `__tests__/session.test.ts` asserts that
 * against a fixture sealed by `core.session.seal_session`.
 *
 * ## One deliberate difference from `@876/core`
 *
 * `unsealSession` rejects a payload with **no** `exp`, matching
 * `core/session.py` (`payload.get("exp", 0) < time.time()`). `@876/core`'s
 * `parsePayload` uses `if (parsed.exp && …)`, so a cookie with no expiry passes
 * there. The stricter reading belongs on the API side: a cookie the API cannot
 * date is a cookie it cannot trust.
 */

import { createHmac, timingSafeEqual } from 'node:crypto'

import { nowUnixSeconds } from '@/platform/timestamps'

/** 400 days — the Python default, and the maximum Chrome honours. */
export const DEFAULT_SESSION_TTL_SECONDS = 60 * 60 * 24 * 400

export type Realm = 'consumer' | 'enterprise'

export type AccountIdentity = {
  userId: string
  email: string
  firstName: string | null
  lastName: string | null
  emailVerified: boolean
  avatar: string | null
  username: string | null
  realm: string
  orgId?: string
  crossRealm?: true
}

export type AccountEntry = AccountIdentity & { sid: string }

export type SessionPayload = AccountIdentity & {
  accessToken: string | null
  exp: number
  sid?: string
  accounts?: AccountEntry[]
}

/**
 * The provider/repository shapes this reads from use either casing, because the
 * Python accepted both — a WorkOS response is camelCase, a database row is
 * snake_case, and both reach this function.
 */
export type UserLike = Record<string, unknown>

function str(value: unknown): string | null {
  return typeof value === 'string' ? value : null
}

function pick(source: UserLike, ...keys: string[]): string | null {
  for (const key of keys) {
    const value = str(source[key])
    if (value) return value
  }
  return null
}

/**
 * Lightweight, token-free identity fields for one account.
 *
 * Shared by the active-account snapshot (the cookie's top level) and by each
 * entry in the multi-account `accounts` list, so the chooser can display every
 * signed-in account and a switch can promote any of them.
 *
 * `crossRealm` is the per-user realm-gate exception, surfaced so the app guards
 * can let an owner or super-admin into any realm. It is carried on every account
 * entry, so switching accounts preserves it.
 */
export function accountIdentity(
  userData: UserLike,
  options: { realm?: string; orgId?: string | null; crossRealm?: boolean } = {}
): AccountIdentity {
  const identity: AccountIdentity = {
    userId: pick(userData, 'id', 'userId') ?? '',
    email: str(userData['email']) ?? '',
    firstName: pick(userData, 'firstName', 'first_name'),
    lastName: pick(userData, 'lastName', 'last_name'),
    emailVerified: Boolean(
      userData['emailVerified'] ?? userData['email_verified'] ?? false
    ),
    avatar: str(userData['avatar']),
    username: str(userData['username']),
    realm: options.realm ?? 'consumer',
  }

  // Both are omitted rather than set to a falsy value, matching the Python —
  // their presence in the payload is itself the signal.
  if (options.orgId) identity.orgId = options.orgId
  if (options.crossRealm) identity.crossRealm = true

  return identity
}

/** One entry in the multi-account `accounts` list — identity plus its sid. */
export function accountEntry(
  userData: UserLike,
  sessionId: string,
  options: { realm?: string; orgId?: string | null; crossRealm?: boolean } = {}
): AccountEntry {
  return { ...accountIdentity(userData, options), sid: sessionId }
}

/**
 * Add `newEntry` to the signed-in account set, de-duped by `userId`.
 *
 * Re-authenticating an account already in the set replaces its stale entry —
 * and its stale sid — rather than appending a duplicate. Order is preserved with
 * the freshly authenticated account last.
 */
export function mergeAccounts(
  existing: AccountEntry[] | null | undefined,
  newEntry: AccountEntry
): AccountEntry[] {
  const kept = (existing ?? []).filter(
    (account) => account.userId !== newEntry.userId
  )

  return [...kept, newEntry]
}

/**
 * The account entry whose `sid` matches, or null when absent.
 *
 * The caller **must** use this to confirm a switch target belongs to the
 * cookie's own account set — never trust a sid supplied by the client.
 */
export function selectAccount(
  accounts: AccountEntry[] | null | undefined,
  sid: string
): AccountEntry | null {
  return (accounts ?? []).find((account) => account.sid === sid) ?? null
}

/**
 * Seal a signed, base64url-encoded session cookie value.
 *
 * The active account's identity stays at the top level, so every existing
 * reader keeps working. When `sessionId` and `accounts` are supplied the cookie
 * additionally carries the active session id and the full signed-in account set
 * for multi-account switching.
 */
export function sealSession(options: {
  userData: UserLike
  accessToken: string | null
  secret: string
  ttlSeconds?: number
  sessionId?: string | null
  accounts?: AccountEntry[] | null
  realm?: string
  orgId?: string | null
  crossRealm?: boolean
}): string {
  const identity = accountIdentity(options.userData, {
    realm: options.realm,
    orgId: options.orgId,
    crossRealm: options.crossRealm,
  })

  const payload: SessionPayload = {
    ...identity,
    accessToken: options.accessToken,
    exp: nowUnixSeconds() + (options.ttlSeconds ?? DEFAULT_SESSION_TTL_SECONDS),
  }
  if (options.sessionId != null) payload.sid = options.sessionId
  if (options.accounts != null) payload.accounts = options.accounts

  const encoded = JSON.stringify(payload)
  const signed = `${encoded}.${sign(encoded, options.secret)}`

  return base64urlEncode(signed)
}

/** Verify the signature and return the payload, or null on any failure. */
export function unsealSession(
  cookieValue: string,
  secret: string
): SessionPayload | null {
  const decoded = base64urlDecode(cookieValue)
  if (decoded === null) return null

  // Split on the *last* dot: the payload is JSON and may contain dots of its own.
  const lastDot = decoded.lastIndexOf('.')
  if (lastDot === -1) return null

  const payloadString = decoded.slice(0, lastDot)
  const signature = decoded.slice(lastDot + 1)

  if (!signaturesMatch(signature, sign(payloadString, secret))) return null

  let payload: unknown
  try {
    payload = JSON.parse(payloadString)
  } catch {
    return null
  }

  if (payload === null || typeof payload !== 'object' || Array.isArray(payload))
    return null

  const candidate = payload as SessionPayload
  // A missing `exp` is treated as expired, not as "never expires".
  if ((candidate.exp ?? 0) < nowUnixSeconds()) return null

  return candidate
}

function sign(payload: string, secret: string): string {
  return createHmac('sha256', secret).update(payload, 'utf8').digest('hex')
}

/**
 * Compare two hex signatures without leaking where they diverge.
 *
 * `timingSafeEqual` throws on a length mismatch, so both sides are hashed to a
 * fixed width first — a forged signature of the wrong length must fail the same
 * way as one of the right length.
 */
function signaturesMatch(a: string, b: string): boolean {
  const left = createHmac('sha256', 'session-signature-compare')
    .update(a)
    .digest()
  const right = createHmac('sha256', 'session-signature-compare')
    .update(b)
    .digest()

  return timingSafeEqual(left, right)
}

/** Unpadded base64url, as the cookie format requires. */
function base64urlEncode(value: string): string {
  return Buffer.from(value, 'utf8').toString('base64url')
}

function base64urlDecode(value: string): string | null {
  try {
    return Buffer.from(value, 'base64url').toString('utf8')
  } catch {
    return null
  }
}
